File: src/main/java/edu/kit/iti/formal/stvs/model/verification/VerificationScenario.java
package edu.kit.iti.formal.stvs.model.verification;

import edu.kit.iti.formal.stvs.logic.io.ExportException;
import edu.kit.iti.formal.stvs.logic.verification.GeTeTaVerificationEngine;
import edu.kit.iti.formal.stvs.logic.verification.VerificationEngine;
import edu.kit.iti.formal.stvs.model.code.Code;
import edu.kit.iti.formal.stvs.model.common.NullableProperty;
import edu.kit.iti.formal.stvs.model.config.GlobalConfig;
import edu.kit.iti.formal.stvs.model.table.ConstraintSpecification;

import java.io.IOException;

import edu.kit.iti.formal.stvs.util.ProcessCreationException;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;

/**
 * The main model object for orchestrating a verification. Has a reference to the currently loaded
 * {@link Code}, uses a {@link VerificationEngine} from the logic package to verify it against a
 * {@link ConstraintSpecification} and provides access to the {@link VerificationResult}.
 *
 * @author Benjamin Alt
 */
public class VerificationScenario {
  private NullableProperty<VerificationResult> verificationResult;
  private VerificationEngine verificationEngine;
  private ObjectProperty<Code> code;
  private ObjectProperty<VerificationState> verificationState;
  private NullableProperty<ConstraintSpecification> activeSpec;

  /**
   * Constructs an empty VerificationScenario (with an empty code).
   */
  public VerificationScenario() {
    this(new Code());
  }

  /**
   * Constructs a VerificationScenario from a given {@link Code}.
   *
   * @param code The initial code
   */
  public VerificationScenario(Code code) {
    this.code = new SimpleObjectProperty<>(code);
    verificationResult = new NullableProperty<>();
    verificationState = new SimpleObjectProperty<>(VerificationState.NOT_STARTED);
    activeSpec = new NullableProperty<>();
  }

  /**
   * Starts a verification of the current {@link Code} against a given
   * {@link ConstraintSpecification}.
   *
   * @param config The config to take into account (i.e. for verification timeouts, paths to
   *        dependencies etc.)
   * @param spec The specification to be verified
   * @throws IOException Exception while IO interaction
   * @throws ExportException Exception while exporting
   */
  public void verify(GlobalConfig config, ConstraintSpecification spec)
      throws IOException, ExportException, ProcessCreationException {
    activeSpec.set(spec);
    verificationEngine =
        new GeTeTaVerificationEngine(config, code.get().getParsedCode().getDefinedTypes());
    verificationEngine.verificationResultProperty().addListener(new VerificationChangedListener());
    verificationState.setValue(VerificationState.RUNNING);
    verificationEngine.startVerification(this, spec);
  }

  /**
   * Stops a currently running verification.
   */
  public void cancel() {
    if (verificationEngine != null) {
      verificationEngine.cancelVerification();
    }
    verificationState.set(VerificationState.CANCELLED);
  }

  public ObjectProperty<VerificationState> verificationState() {
    return verificationState;
  }

  public Code getCode() {
    return code.get();
  }

  public void setCode(Code code) {
    this.code.set(code);

  }

  public ConstraintSpecification getActiveSpec() {
    return activeSpec.get();
  }

  public void setActiveSpec(ConstraintSpecification spec) {
    activeSpec.set(spec);
  }

  public NullableProperty<ConstraintSpecification> activeSpecProperty() {
    return activeSpec;
  }

  public ObjectProperty<Code> codeObjectProperty() {
    return this.code;
  }

  public VerificationResult getVerificationResult() {
    return verificationResult.get();
  }

  public NullableProperty<VerificationResult> verificationResultProperty() {
    return verificationResult;
  }

  private class VerificationChangedListener implements ChangeListener<VerificationResult> {
    @Override
    public void changed(ObservableValue<? extends VerificationResult> observableValue,
        VerificationResult oldResult, VerificationResult newResult) {
      verificationResult.set(newResult);
      verificationState.set(VerificationState.FINISHED);
    }
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/model/verification/VerificationScenario.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.model.verification

import edu.kit.iti.formal.stvs.logic.io.ExportException
import edu.kit.iti.formal.stvs.logic.verification.GeTeTaVerificationEngine
import edu.kit.iti.formal.stvs.logic.verification.VerificationEngine
import edu.kit.iti.formal.stvs.model.code.Code
import edu.kit.iti.formal.stvs.model.common.NullableProperty
import edu.kit.iti.formal.stvs.model.config.GlobalConfig
import edu.kit.iti.formal.stvs.model.table.ConstraintSpecification
import edu.kit.iti.formal.stvs.util.ProcessCreationException
import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableValue
import tornadofx.getValue
import tornadofx.setValue
import java.io.IOException

/**
 * The main model object for orchestrating a verification. Has a reference to the currently loaded
 * [Code], uses a [VerificationEngine] from the logic package to verify it against a
 * [ConstraintSpecification] and provides access to the [VerificationResult].
 *
 * @author Benjamin Alt
 */
class VerificationScenario {
    val verificationResultProperty = NullableProperty<VerificationResult>()
    var verificationResult by verificationResultProperty

    var verificationEngine: VerificationEngine? = null

    val codeProperty: ObjectProperty<Code> = SimpleObjectProperty(Code())
    var code: Code by codeProperty

    val verificationStateProperty = SimpleObjectProperty(VerificationState.NOT_STARTED)
    val verificationState by verificationStateProperty

    val activeSpecProperty = NullableProperty<ConstraintSpecification>()
    var activeSpec: ConstraintSpecification? by activeSpecProperty

    /**
     * Constructs a VerificationScenario from a given [Code].
     *
     * @param code The initial code
     */

    /**
     * Starts a verification of the current [Code] against a given
     * [ConstraintSpecification].
     *
     * @param config The config to take into account (i.e. for verification timeouts, paths to
     * dependencies etc.)
     * @param spec The specification to be verified
     * @throws IOException Exception while IO interaction
     * @throws ExportException Exception while exporting
     * @throws ProcessCreationException exception while creating process
     */
    @Throws(IOException::class, ExportException::class, ProcessCreationException::class)
    fun verify(config: GlobalConfig, spec: ConstraintSpecification) {
        activeSpec = spec
        val ve = GeTeTaVerificationEngine(config, codeProperty.get().parsedCode!!)
        ve.verificationResultProperty.addListener(VerificationChangedListener())
        verificationStateProperty.value = VerificationState.RUNNING
        ve.startVerification(this, ConstraintSpecification(spec))

        verificationEngine = ve
    }

    /**
     * Stops a currently running verification.
     */
    fun cancel() {
        if (verificationEngine != null) {
            verificationEngine!!.cancelVerification()
        }
        verificationStateProperty.set(VerificationState.CANCELLED)
    }

    private inner class VerificationChangedListener : ChangeListener<VerificationResult?> {
        override fun changed(
            observable: ObservableValue<out VerificationResult?>?,
            oldValue: VerificationResult?,
            newValue: VerificationResult?,
        ) {
            verificationResult = newValue
            verificationStateProperty.set(VerificationState.FINISHED)
        }
    }
}